### ISSUE TITLE
run erpc server task in multi threaded environment

### ISSUE THREAD

**Issue by tmordeko** [EmbeddedRPC/erpc#95]
Hi,

Running with FreeRTOS OS, erpc_server_run() is called from from separate task.
The way previous function is implemented is in endless loop, never giving up on cpu waiting for ever for received data on the transport layer.

Is there a way to optimize that code such that it will be waiting on events from the transport layer ?

Thanks,
Tzvika

**Comment by Hadatko** [comment]
Hi @tmordeko , di you consider poll function instead of run?

**Comment by tmordeko** [comment]
Hi @Hadatko,

I am aware of the poll mode, but it basically brings me back to the same problem as I am waiting on the uart channel.

Tzvika

**Comment by Hadatko** [comment]
I see. Looks like you need another implementation of uart receive function. For example erpc could have possibility to have ring buffers and save messages on background (something like rpmsg).

**Comment by tmordeko** [comment]
is it something supported today or I should implement it on my own ?

**Comment by MichalPrincNXP** [comment]
Hi @tmordeko , have you coped with this problem? Any update?

**Comment by Hadatko** [comment]
@tmordeko today i was talking with @MichalPrincNXP about this issue. It looks like solution is simple to add erpc_thread.h and implement take/put sempahores based on that header file. Take semaphore instead of while loop and put semaphore from interrupt once data copied. Do you think you can implement this feature these days? If we will do it quickly it will show up in release next release (1.7.5?).

**Comment by tmordeko** [comment]
Hi @Hadatko ,

Thanks for the update, 

When is next release1.7.5 expected to be published ?

Can you do the change and send me the CR so we will integrate it ?

/Tzvika

**Comment by Hadatko** [comment]
If you are not interested in doing change i will let it on @MichalPrincNXP.

**Comment by MichalPrincNXP** [comment]
Hello, this seems to me like important and easy to fix issue. I would try to do my best to update the uart transport layer by the end of next week when the internal code freeze date is planned. Will provide you the patch later, ok?

**Comment by Hadatko** [comment]
@MichalPrincNXP just thinking what should be best way of developing this framework from NXP site. I think the new features could be synced back to develop on github once they are pushed into internal repository. So community can see progress and what will be in next release and they will have also access to the latest code.

If we would have bigger active users, then NXP could also do PR to github to get review from community. Just thinking....

**Comment by tmordeko** [comment]
Hi @MichalPrincNXP 

Thanks a bunch !

**Comment by MichalPrincNXP** [comment]
> 
> 
> @MichalPrincNXP just thinking what should be best way of developing this framework from NXP site. I think the new features could be synced back to develop on github once they are pushed into internal repository. So community can see progress and what will be in next release and they will have also access to the latest code.
> 
> If we would have bigger active users, then NXP could also do PR to github to get review from community. Just thinking....

yes, agree, we should get back to the monthly update concept, but ... it is not easy to manage all that ...

**Comment by MichalPrincNXP** [comment]
Hello @tmordeko ,
attached is the updated CMSIS UART transport layer ... could you please try on your side and let me know if it helps to solve your problem? I have made quick testing on an erpc example that runs FreeRTOS-based client on the embedded side and the erpc server on the PC side, using this updated CMSIS UART transport, and it seems to be working fine. Thanks.

Note1: erpc_threading.h and erpc_threading_freertos.cpp needs to be added into your project if not already done.
Note2: UART interrupt service priority needs to be set correctly in FreeRTOS env. otherwise assert could be reached (if enabled), see portASSERT_IF_INTERRUPT_PRIORITY_INVALID. One could use the NVIC_SetPriority() function for this purpose, for instance: NVIC_SetPriority(FLEXCOMM0_IRQn, 5);
[transports.zip](https://github.com/EmbeddedRPC/erpc/files/5316576/transports.zip)

**Comment by Hadatko** [comment]
@MichalPrincNXP it would be fine in future to have PR instead of zip file. It is better way to comment changes from various od peaople and easy merge process + history...

**Comment by Hadatko** [comment]
I think this is already merged, am i correct? Can be this done also for other transports like e.g.: MU???

**Comment by MichalPrincNXP** [comment]
Yes, it has been integrated already, thanks for the notice, I am closing now ...
